src/main/java/frc/robot/vision/sources/limelights/LimeLightSource.java

```diff
@@ -0,0 +1,193 @@
+				gyroAngleValues.rollRate()}
+		);
+		robotPoseArray = (useGyroForPoseEstimating ? robotPoseEntryBotPose2 : robotPoseEntryBotPose1)
+			.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		aprilTagPoseArray = aprilTagPoseEntry.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		robotPoseWithoutGyroInput = robotPoseEntryBotPose1.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		robotHeading = Rotation2d.fromDegrees(robotPoseWithoutGyroInput[LimelightEntryIndex.YAW_ANGLE.getIndex()]);
+		standardDeviationsArray = standardDeviations.getDoubleArray(new double[Pose3dComponentsValue.POSE3D_COMPONENTS_AMOUNT]);
+
+		log();
+	}
+
+	private Optional<Pair<Pose3d, Double>> getUpdatedPose3DEstimation() {
+		int id = (int) aprilTagIdEntry.getInteger(VisionConstants.NO_APRILTAG_ID);
+		if (id == VisionConstants.NO_APRILTAG_ID) {
+			return Optional.empty();
+		}
+
+		double processingLatencySeconds = Conversions.milliSecondsToSeconds(robotPoseArray[LimelightEntryIndex.TOTAL_LATENCY.getIndex()]);
+		double timestamp = TimeUtils.getCurrentTimeSeconds() - processingLatencySeconds;
+
+		Pose3d robotPose = new Pose3d(
+			getPoseValue(LimelightEntryIndex.X_AXIS),
+			getPoseValue(LimelightEntryIndex.Y_AXIS),
+			getPoseValue(LimelightEntryIndex.Z_AXIS),
+			new Rotation3d(
+				Math.toRadians(getPoseValue(LimelightEntryIndex.ROLL_ANGLE)),
+				Math.toRadians(getPoseValue(LimelightEntryIndex.PITCH_ANGLE)),
+				Math.toRadians(getPoseValue(LimelightEntryIndex.YAW_ANGLE))
+			)
+		);
+		return Optional.of(new Pair<>(robotPose, timestamp));
+	}
+
+	public double getPoseValue(LimelightEntryIndex entryValue) {
+		return robotPoseArray[entryValue.getIndex()];
+	}
+
+	public double getAprilTagValueInRobotSpace(LimelightEntryIndex entryValue) {
+		return aprilTagPoseArray[entryValue.getIndex()];
+	}
+
+	@Override
+	public Optional<AprilTagVisionData> getVisionData() {
+		Optional<Pair<Pose3d, Double>> poseEstimation = getUpdatedPose3DEstimation();
+		return poseEstimation.map(
+			pose3dDoublePair -> new AprilTagVisionData(
+				name,
+				pose3dDoublePair.getFirst(),
+				pose3dDoublePair.getSecond(),
+				standardDeviationsArray,
+				getAprilTagValueInRobotSpace(LimelightEntryIndex.Z_AXIS),
+				getAprilTagValueInRobotSpace(LimelightEntryIndex.Y_AXIS),
+				(int) aprilTagIdEntry.getInteger(VisionConstants.NO_APRILTAG_ID) // a safe cast as long as limelight doesn't break APIs
+			)
+		);
+	}
+
+	@Override
+	public Optional<AprilTagVisionData> getFilteredVisionData() {
+		if (shouldDataBeFiltered(getVisionData())) {
+			return getVisionData();
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	private boolean shouldDataBeFiltered(Optional<AprilTagVisionData> data) {
+		return data.map(filter::apply).orElseGet(() -> true);
+	}
+
+	@Override
+	public Filter<AprilTagVisionData> setFilter(Filter<AprilTagVisionData> newFilter) {
+		return this.filter = newFilter;
+	}
```

Comment:
maybe add: addFilter() that is not deleting the filters u curretlly have.
Also maybe add clearFilters that removes all filters

Reply:
Right; but before it's done here, you should add that comment to Yotam's PR (#519) because I don't want to modify the interface `VisionSource` here. I'll keep that open till this function will be implemented here too.

Reply:
Adding a filter is not a clear operation as there are two ways to combine filters (`and`, `or`), do you want two methods instead?